### PR TITLE
refactor(influxql): remove `copy_with_replacement` and use upstream API

### PR DESCRIPTION
**Pull request by alamb**
# Rationale

The influxql crate has a copy/paste version of `copy_with_replacement` which can be implemented with the DataFusion [TreeNode API](https://docs.rs/datafusion/latest/datafusion/common/tree_node/index.html)

This code is annoying to update as the `Expr` structure evolves in DataFusion (e.g. https://github.com/influxdata/influxdb_iox/pull/8447#discussion_r1286332552) and it turns out it isn't needed in DataFusion either: https://github.com/apache/arrow-datafusion/pull/7232

# Changes
Remove `copy_with_replacement` and use TreeNode API

### CODE REVIEW

File: iox_query_influxql/src/plan/planner.rs [hunk 1]
```diff
@@ -1186,12 +1185,13 @@ impl<'a> InfluxQLToLogicalPlan<'a> {
 
                     udf.args.append(&mut additional_args);
                     let selector_new = Expr::AggregateUDF(udf);
-                    select_exprs[selector_index] =
-                        clone_with_replacement(&select_exprs[selector_index], &|expr| {
-                            if expr == &selector {
-                                Ok(Some(selector_new.clone()))
+                    select_exprs[selector_index] = select_exprs[selector_index]
```
Comment (alamb) [inline iox_query_influxql/src/plan/planner.rs:1188]: here is how the TreeNode API is used

File: iox_query_influxql/src/plan/util.rs [hunk 2]
```diff
@@ -186,3 +185,53 @@ pub(crate) fn contains_expr(expr: &Expr, needle: &Expr) -> bool {
...
     found
 }
+
+/// Search the provided `Expr`'s, and all of their nested `Expr`, for any that
+/// pass the provided test. The returned `Expr`'s are deduplicated and returned
+/// in order of appearance (depth first).
+///
+/// # NOTE
+///
+/// Copied from DataFusion
```
Comment (alamb) [inline iox_query_influxql/src/plan/util.rs:195]: I moved this from `util_copy` as it has nothing to do with copying

File: iox_query_influxql/src/plan/util_copy.rs [hunk 3]
```diff
@@ -1,390 +0,0 @@
...
-/// * `Ok(Some(replacement_expr))`: A replacement `Expr` is provided; it is
-///       swapped in at the particular node in the tree. Any nested `Expr` are
-///       not subject to cloning/replacement.
-/// * `Ok(None)`: A replacement `Expr` is not provided. The `Expr` is
-///       recreated, with all of its nested `Expr`'s subject to
-///       cloning/replacement.
-/// * `Err(err)`: Any error returned by the function is returned as-is by
-///       `clone_with_replacement()`.
-///
-/// Source: <https://github.com/apache/arrow-datafusion/blob/26e1b20ea/datafusion/sql/src/utils.rs#L153>
```
Comment (alamb) [inline iox_query_influxql/src/plan/util_copy.rs:44]: I propose removing this upstream too: https://github.com/apache/arrow-datafusion/pull/7232

**Review by mhilton — APPROVED** [review]
LGTM, nice cleanup thanks

File: iox_query_influxql/src/plan/util.rs (hunk 2 above)
Comment (stuartcarnie) [reply iox_query_influxql/src/plan/util.rs:195]: heh, the `copy` suffix only meant the code in this file was copied from DataFusion, and hopefully temporary, when the relevant functions became available or unnecessary.

Really nice cleanup, thanks!

Comment (alamb) [reply iox_query_influxql/src/plan/util.rs:195]: Makes sense -- thank you@

**Pull request merged**